File: das_custom_template.m
function [bmode_img, delays] = das_custom_template(rf_data, tx_info, rx_pos, grid_x, grid_z, sound_speed, fs)
%DAS_CUSTOM_TEMPLATE  Skeleton for writing your own DAS beamformer.
%
%   [bmode_img, delays] = DAS_CUSTOM_TEMPLATE(rf_data, tx_info, rx_pos, ...
%                                             grid_x, grid_z, sound_speed, fs)
%
%   HOW TO USE
%     1) Copy this file under a name of your choice (e.g. my_das.m).
%     2) Rename the function so it matches the file name.
%     3) Edit [STEP 1] ... [STEP 5] below.
%     4) Type that name into the "Custom function" field of MAIN_GUI and pick
%        "Custom DAS" as algorithm A or B to compare it with the reference.
%
%   CONTRACT (breaking it breaks the comparison and the metrics)
%     - Keep the argument order and meaning identical to DAS_REFERENCE.
%     - rf_data is [Nt x Nel] (time along the first dimension, columns are
%       receive elements).
%     - bmode_img must be [numel(grid_z) x numel(grid_x)] and must be the
%       **linear envelope** (no log compression). A wrong size makes the
%       difference image and the MSE fail.
%     - delays must be [Npix x Nel] holding the **total two-way time [s]**,
%       with NaN wherever a sample is not summed. Skip it when nargout < 2.
%     - Pixels are ordered column-major (the linear index after reshape).
%
%   INITIAL STATE
%     A deliberately naive DAS (nearest-neighbour sampling plus a rectangular
%     apodisation) so the file runs out of the box. Comparing it with the
%     reference (linear interpolation) shows the interpolation error as a
%     raised sidelobe floor in the difference image and the lateral profile.
%
%   See also DAS_REFERENCE, SIM_ENGINE, MAIN_GUI.

narginchk(7, 7);

%% =====================================================================
%  [STEP 0] Input handling (usually left untouched)
%% =====================================================================
if ndims(rf_data) > 2 %#ok<ISMAT>
    rf_data = rf_data(:,:,1);
end
[Nt, Nel] = size(rf_data);

rxx = rx_pos(:).';                       % receive element x coordinates [m]
rxz = zeros(1, Nel);                     % linear array, so z = 0
assert(numel(rxx) == Nel, 'rx_pos does not match the number of RF columns.');

if isvector(grid_x) && isvector(grid_z)
    [XI, ZI] = meshgrid(grid_x(:).', grid_z(:).');
else
    XI = grid_x; ZI = grid_z;
end
imgSize = size(XI);
xg = XI(:); zg = ZI(:); Npix = numel(xg);

c  = sound_speed;
t0 = tx_info.t0;                         % time of the first RF sample [s]

%% =====================================================================
%  [STEP 1] Transmit arrival time tau_tx(p)   <<< EDIT HERE
%  ---------------------------------------------------------------------
%  Two models are provided, mirroring the reference implementation.
%   (a) First-arrival (Huygens) model - exact for plane, diverging and
%       single-element transmits:
%         tau_tx(p) = min_e ( delays(e) + |p - elem_e| / c )
%   (b) Virtual-source model - used for focused transmits:
%         tau_tx(p) = T_F + s * |p - F| / c   (s = -1 before, +1 beyond focus)
%  Using (a) for a focused transmit makes the beamformer latch onto the weak
%  edge wave beyond the focus and shifts the image by roughly 1 mm in depth.
%  Removing the distinction is an instructive experiment in itself.
%% =====================================================================
ex  = tx_info.elem_x;
ez  = tx_info.elem_z;
td  = tx_info.delays;
ap  = tx_info.apod;
act = isfinite(td) & ap(:).' ~= 0;       % only the elements that actually fire
exA = ex(act); ezA = ez(act); tdA = td(act);

src = tx_info.src_xz;                    % focal point / virtual source [x z], NaN if none
useVirtualSource = strcmpi(tx_info.scheme,'focused') && all(isfinite(src)) && ...
                   src(2) > 0 && ~tx_info.single_element;

if useVirtualSource
    Rf   = hypot(exA - src(1), ezA - src(2));
    Tf   = mean(tdA + Rf / c);           % instant the wavefront converges on the focus [s]
    D    = hypot(src(1), src(2));
    u    = [src(1) src(2)] / D;
    proj = xg*u(1) + zg*u(2);
    dF   = hypot(xg - src(1), zg - src(2));
    sgn  = ones(Npix,1);  sgn(proj <= D) = -1;
    tau_tx = Tf + sgn .* dF / c;
else
    tau_tx = inf(Npix, 1);
    for e = 1:numel(exA)
        tau_tx = min(tau_tx, tdA(e) + hypot(xg - exA(e), zg - ezA(e)) / c);
    end
end

%% =====================================================================
%  [STEP 2] Receive aperture weights (f-number, apodisation)  <<< EDITABLE
%% =====================================================================
fnum = tx_info.fnumber;
if ~isfinite(fnum) || fnum <= 0
    halfAp = inf(Npix, 1);               % full aperture
else
    halfAp = zg ./ (2*fnum);             % aperture grows with depth
end

%% =====================================================================
%  [STEP 3] Delay-and-sum main loop   <<< EDIT HERE
%  ---------------------------------------------------------------------
%  The initial version uses nearest-neighbour sampling. Replacing round()
%  with floor() plus linear interpolation should reproduce the reference
%  implementation exactly, which makes a handy self-check.
%% =====================================================================
bf = zeros(Npix, 1);
if nargout > 1
    delays = nan(Npix, Nel);
end

for e = 1:Nel
    dx  = xg - rxx(e);
    tau = tau_tx + hypot(dx, zg - rxz(e)) / c;    % total two-way time [s]

    w   = double(abs(dx) <= halfAp);              % rectangular apodisation

    idx = round((tau - t0) * fs + 1);             % <-- nearest-neighbour sampling
    ok  = w > 0 & idx >= 1 & idx <= Nt;
    idxc = min(max(idx, 1), Nt);

    val = rf_data(idxc, e);
    val(~ok) = 0;

    bf = bf + w .* val;

    if nargout > 1
        d = tau; d(~ok) = NaN;
        delays(:, e) = d;
    end
end

%% =====================================================================
%  [STEP 4] Envelope detection   <<< EDITABLE (try something other than Hilbert)
%% =====================================================================
bf2 = reshape(bf, imgSize);
if imgSize(1) >= 4
    bmode_img = abs(hilbert(bf2));
else
    bmode_img = abs(bf2);
end

%% =====================================================================
%  [STEP 5] Post-processing (TGC, speckle reduction, ...) goes here
%% =====================================================================
% bmode_img = ...;

end

File: python/ultrasound_das/das_custom_template.py
"""Skeleton for writing your own DAS beamformer (Python port of
``das_custom_template.m``).

HOW TO USE
    1) Copy this file, e.g. to ``my_das.py``.
    2) Rename ``das_custom_template`` to ``my_das``.
    3) In the GUI (Detailed mode) type ``my_das`` into "Custom function name"
       and pick "Custom DAS" as algorithm A or B, or pass the callable to
       :func:`ultrasound_das.pipeline.run_algorithm`.

CONTRACT (identical to :func:`ultrasound_das.das_reference.das_reference`)
    - Keep the argument order and meaning.
    - ``rf_data`` is [Nt x Nel] (time along the first dimension).
    - ``bmode_img`` must be [len(grid_z) x len(grid_x)] and the **linear
      envelope** (no log compression).
    - ``delays`` must be [Npix x Nel] holding the total two-way time [s], NaN
      where a sample is not summed. Pixels are column-major
      (``img.ravel(order='F')``). Only build it when ``want_delays=True``.

INITIAL STATE
    A deliberately naive DAS (nearest-neighbour sampling + rectangular
    apodisation) so the file runs out of the box. Comparing it with the
    reference (linear interpolation) shows the interpolation error as a raised
    sidelobe floor in the difference image and the lateral profile.
"""

# want_trace=True returns (image, delays, trace) with actual channel increments.
# Keep the accumulation hook in custom implementations; see docs/HW3_DAS.md.

from __future__ import annotations

import numpy as np


def _get(tx, name, default):
    v = tx.get(name, default) if isinstance(tx, dict) else getattr(tx, name, default)
    return default if v is None else v


def das_custom_template(
    rf_data, tx_info, rx_pos, grid_x, grid_z, sound_speed, fs, want_delays: bool = False, want_trace: bool = False
):
    # ---- [STEP 0] Input handling (usually left untouched) ----
    rf_data = np.asarray(rf_data, dtype=float)
    if rf_data.ndim > 2:
        rf_data = rf_data[:, :, 0]
    nt, nel = rf_data.shape

    rxx = np.asarray(rx_pos, dtype=float).ravel()
    rxz = np.zeros(nel)
    if rxx.size != nel:
        raise ValueError("rx_pos does not match the number of RF columns.")

    gx = np.atleast_1d(np.asarray(grid_x, dtype=float))
    gz = np.atleast_1d(np.asarray(grid_z, dtype=float))
    if gx.ndim == 1 and gz.ndim == 1:
        XI, ZI = np.meshgrid(gx, gz)
    else:
        XI, ZI = gx, gz
    img_size = XI.shape
    xg = XI.ravel(order="F")
    zg = ZI.ravel(order="F")
    npix = xg.size

    c = float(sound_speed)
    t0 = float(_get(tx_info, "t0", 0.0))

    # ---- [STEP 1] Transmit arrival time tau_tx(p)   <<< EDIT HERE ----
    ex = np.asarray(_get(tx_info, "elem_x", None), dtype=float)
    ez = np.asarray(_get(tx_info, "elem_z", np.zeros_like(ex)), dtype=float)
    td = np.asarray(_get(tx_info, "delays", None), dtype=float)
    ap = np.asarray(_get(tx_info, "apod", np.ones_like(ex)), dtype=float)
    act = np.isfinite(td) & (ap.ravel() != 0)
    exA, ezA, tdA = ex[act], ez[act], td[act]

    src = np.asarray(_get(tx_info, "src_xz", (np.nan, np.nan)), dtype=float)
    use_virtual_source = (
        str(_get(tx_info, "scheme", "")).lower() == "focused"
        and np.all(np.isfinite(src))
        and src[1] > 0
        and not bool(_get(tx_info, "single_element", False))
    )

    if use_virtual_source:
        Rf = np.hypot(exA - src[0], ezA - src[1])
        Tf = np.mean(tdA + Rf / c)
        D = np.hypot(src[0], src[1])
        u = src / D
        proj = xg * u[0] + zg * u[1]
        dF = np.hypot(xg - src[0], zg - src[1])
        sgn = np.ones(npix)
        sgn[proj <= D] = -1.0
        tau_tx = Tf + sgn * dF / c
    else:
        tau_tx = np.full(npix, np.inf)
        for e in range(exA.size):
            tau_tx = np.minimum(tau_tx, tdA[e] + np.hypot(xg - exA[e], zg - ezA[e]) / c)

    # ---- [STEP 2] Receive aperture weights (f-number, apodisation) ----
    fnum = float(_get(tx_info, "fnumber", 0.0))
    if not np.isfinite(fnum) or fnum <= 0:
        half_ap = np.full(npix, np.inf)
    else:
        half_ap = zg / (2.0 * fnum)

    # ---- [STEP 3] Delay-and-sum main loop   <<< EDIT HERE ----
    #  Nearest-neighbour sampling. Replacing round() with floor() + linear
    #  interpolation should reproduce the reference implementation exactly.
    bf = np.zeros(npix)
    delays = np.full((npix, nel), np.nan) if (want_delays or want_trace) else None

    contributions = np.zeros((npix, nel)) if want_trace else None

    for e in range(nel):
        dx = xg - rxx[e]
        tau = tau_tx + np.hypot(dx, zg - rxz[e]) / c
        w = (np.abs(dx) <= half_ap).astype(float)

        # nearest-neighbour, 0-based. floor(x + 0.5) matches MATLAB round()
        # (round-half-away-from-zero) for the non-negative sample positions here.
        idx = np.floor((tau - t0) * fs + 0.5).astype(np.int64)
        ok = (w > 0) & (idx >= 0) & (idx < nt)
        idxc = np.clip(idx, 0, nt - 1)
        val = np.where(ok, rf_data[idxc, e], 0.0)

        previous = bf.copy() if want_trace else None
        bf += w * val
        if want_trace:
            contributions[:, e] = bf - previous
        if want_delays or want_trace:
            d = tau.copy()
            d[~ok] = np.nan
            delays[:, e] = d

    # ---- [STEP 4] Envelope detection   <<< EDITABLE ----
    bf2 = bf.reshape(img_size, order="F")
    if img_size[0] >= 4:
        from .das_reference import envelope_z

        bmode_img = envelope_z(bf2)
    else:
        bmode_img = np.abs(bf2)

    # ---- [STEP 5] Post-processing (TGC, speckle reduction, ...) ----

    if want_trace:
        return bmode_img, delays, {"contributions": contributions,
                                   "coherent": bf.reshape(img_size, order="F")}
    if want_delays:
        return bmode_img, delays
    return bmode_img
